### PR TITLE
Add mirror copy feature

### PR DESCRIPTION
**Pull request by iPersona**
see #5

### CODE REVIEW

**Comment by webdesus** [comment]
Thanks man. I look this PR in near time😊

**Comment by iPersona** [comment]
I forget to remove the debug log :stuck_out_tongue_winking_eye: , fix here e696caebb78c6dde7d9dc1e6a1924a19c604bdea

File: tests/dir.rs [hunk 1]
```diff
@@ -1070,8 +1070,84 @@ fn it_copy_with_progress_exist_overwrite_and_skip_exist() {
 
 }
 
+#[test]
+fn it_copy_mirror_work() {
```
Comment (webdesus) [inline tests/dir.rs]: One test not enough for this operation. I see 3 situation( command for examples will `cp -r d1 d2`)

**First**

d2 not exist 

``` tree.
.
└── d1
   ├── f1
   └── sub
      └── f2
```
*result*
``` tree.
.
└── d1
   ├── f1
   └── sub
      └── f2
└── d2
   ├── f1
   └── sub
      └── f2

```


**Second**

d2 exist but not contains d1 folder

``` tree.
.
└── d1
   ├── f1
   └── sub
      └── f2
└── d2
      └── fzz
```
*result*
``` tree.
.
└── d1
   ├── f1
   └── sub
      └── f2
└── d2
    └── d1
       ├── f1
       └── sub
          └── f2
    └── fzz

```

**Third**

d2 exist and contains some files of d1 folder

``` tree.
.
└── d1
   ├── f1
   └── sub
      └── f2
└── d2
    └── d1
       ├── f1
       └── sub
          └── f2

```
*result*
``` tree.
.
└── d1
   ├── f1
   └── sub
      └── f2
└── d2
    └── d1
       ├── f1   // will be overwrite if will be set flag overwrite
       └── sub
          └── f2 // will be overwrite if will be set flag overwrite
    └── fzz

```

Comment (iPersona) [reply tests/dir.rs]: Hi, webdesus. I think the situation 2&3 can be simply done by checking d2 folder and and it's sub folder and then determine whether to copy to the sub folder. 
Maybe I just implement the cp command with a rename feature as situation 1. :blush:

File: tests/dir.rs [hunk 2]
```diff
@@ -1070,8 +1070,84 @@ fn it_copy_with_progress_exist_overwrite_and_skip_exist() {
...
+    options.overwrite = true;
+    let result = copy(&path_from, &path_to, &options).unwrap();
 
+    assert_eq!(16, result);
+    assert!(path_to.exists());
+    assert!(path_from.exists());
+    assert!(compare_dir_mirror(&path_from, &path_to));
+}
+
+fn compare_dir_mirror<P, Q>(path_from: P, path_to: Q) -> bool
```
[the remark this answers is not in the document]
Comment (iPersona) [reply tests/dir.rs]: The original function compare_dir seems to be working when the two folder to be compared with the same sub folder name. when comparing with folder as situation 1, it does not work correctly. So I write a new method to make it working in situation 1.

File: src/dir.rs [hunk 3]
```diff
@@ -13,6 +13,8 @@ pub struct CopyOptions {
     pub skip_exist: bool,
     /// Sets buffer size for copy/move work only with receipt information about process work.
     pub buffer_size: usize,
+    /// Sets the option true for mirroring directory structure when copying directory.
+    pub mirror_copy: bool,
```
[the remark this answers is not in the document]
Comment (iPersona) [reply src/dir.rs]: yeah, the copy_inside may be better, and set to true by default to make the previous function work as it used to be. :blush:

File: tests/dir.rs (hunk 1 above)
Comment (webdesus) [reply tests/dir.rs]: I think better will be write test before:wink:

File: tests/dir.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (webdesus) [reply tests/dir.rs]: i did not see difference)

[the remark this answers is not in the document]
Comment (iPersona) [reply tests/dir.rs]: The `compare_dir` method assumes that the folder `path_to` must have a sub folder named the last component of the `path_from`. So in order to compare two directories with the same structure but different name, I rewrite a new directory comparing method named `compare_dir_recursively` :smile:

File: tests/dir.rs (hunk 1 above)
Comment (iPersona) [reply tests/dir.rs]: Three test cases add!

**Comment by webdesus** [comment]
ok i watch it today or tomorrow😊

**Comment by webdesus** [comment]
I will watch code `copy` soon. But for good terms needed make this options actual for 

- move_dir
- copy_with_progress
- move_dir_with_progress

**Comment by iPersona** [comment]
ok, I will see that later in few days :blush:

**Comment by iPersona** [comment]
Hi webdesus, I had finish the three feature as you mansioned ☺️:
- move_dir
- copy_with_progress
- move_dir_with_progress

**Comment by webdesus** [comment]
Big thanks. I check soon 😀

**Comment by iPersona** [comment]
Thanks. I'm looking forward to it. :grin:

**Comment by iPersona** [comment]
Hi webdesus, I had removed the unnecessary code here eb9a3f2bfc3c72e010fcad0a9e0e94170b64f21b 😁

**Comment by webdesus** [comment]
Thanks, but you code have one problem `let mut to = to.to_path_buf();` you should removed `mut`  [link](https://github.com/webdesus/fs_extra/pull/6/commits/eb9a3f2bfc3c72e010fcad0a9e0e94170b64f21b#diff-c1f67c4980f2124dc5ea1cb8d2b99eedL529)

**Comment by iPersona** [comment]
Sorry, my fault 😁 . new fix here 402ba0bf28157100b72a25935daa9444fc4716f7

**Comment by webdesus** [comment]
Thank you😁 You very help me😁

**Comment by iPersona** [comment]
You are welcome :grind:

**Pull request merged**